Sources/GRPCCore/GRPCClient.swift

```diff
@@ -0,0 +1,823 @@
+/// This gives the client enough time to drain any requests already in flight. To stop the client more abruptly
+/// you can cancel the task running your client. If your application requires additional resources
+/// that need their lifecycles managed you should consider using [Swift Service
+/// Lifecycle](https://github.com/swift-server/swift-service-lifecycle).
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+public final class GRPCClient: Sendable {
+  /// A collection of ``ClientInterceptor`` implementations which are applied to all accepted
+  /// RPCs.
+  ///
+  /// RPCs are intercepted in the order that interceptors are added. That is, a request sent from the client to
+  /// the server will first be intercepted by the first added interceptor followed by the second, and so on.
+  /// For responses from the server, they'll be applied in the opposite order.
+  public var interceptors: Interceptors {
+    get {
+      self.storage.withLockedValue { $0.interceptors }
+    }
+    set {
+      self.storage.withLockedValue { storage in
+        if case .notStarted = storage.state {
+          storage.interceptors = newValue
+        }
+      }
+    }
+  }
+
+  /// A ``MethodConfigurationRegistry`` containing ``MethodConfiguration``s for calls
+  /// made from this ``Client``.
+  ///
+  /// - Note: These configurations will override those configurations set in the ``ClientTransport``.
+  public var methodConfigurationOverrides: MethodConfigurationRegistry {
+    get {
+      self.storage.withLockedValue { $0.methodConfigurationOverrides }
+    }
+    set {
+      self.storage.withLockedValue { storage in
+        if case .notStarted = storage.state {
+          storage.methodConfigurationOverrides = newValue
+        }
+      }
+    }
+  }
+
+  /// The state of the client.
+  private enum State {
+    /// The client hasn't been started yet. Can transition to `running` or `stopped`.
+    case notStarted
+    /// The client is running and can send RPCs. Can transition to `stopping`.
+    case running
+    /// The client is stopping and no new RPCs will be sent. Existing RPCs may run to
+    /// completion. May transition to `stopped`.
+    case stopping
+    /// The client has stopped, no RPCs are in flight and no more will be accepted. This state
+    /// is terminal.
+    case stopped
+  }
+
+  /// Underlying storage for the client.
+  private struct Storage {
+    var interceptors: Interceptors
+    var methodConfigurationOverrides: MethodConfigurationRegistry
+    var state: State
+
+    init() {
+      self.interceptors = Interceptors()
+      self.methodConfigurationOverrides = MethodConfigurationRegistry()
+      self.state = .notStarted
+    }
+  }
+
+  private let storage: LockedValueBox<Storage>
+
+  /// The transport which provides a bidirectional communication channel with the server.
+  private let transport: ClientTransport
+
+  /// Creates a new client with no resources.
```

Comment:
I find `resources` a bit weird naming wise. Like what resource?

Reply:
I borrowed the naming from the docs of the `GRPCServer`. In this case, it refers to the `Interceptors` and the `MethodConfiguration`s. Happy to change this to a different word, or add a "(i.e.: `Interceptors` and `MethodConfiguration`s)" to make it clearer :D